classes/patreon_login.php

```diff
@@ -14,72 +14,208 @@ public static function updateExistingUser($user_id, $user_response, $tokens) {
+	public static function checkTokenExpiration($user_id=false) {
+
+		if($user_id)
+		{
+			$user = get_user_by('ID',$user_id);
+		}
+		else
+		{
+			$user = wp_get_current_user();
+		}
+
+		if(!$user OR 0 == $user->ID) {
+		} else {
+			
+			// Valid user is logged in. Check the token:
+			
+			$expiration = get_user_meta($user->ID,'patreon_token_expires_in',true);
+			$minted = get_user_meta($user->ID,'patreon_token_minted',true);
+			
+			if($minted!='')
+			{
+				// We have value. get secs to use them in comparison.
+
+				$minted = explode(' ',$minted);
+				// Cast to integer
+				$minted = (int) $minted[1];
+				
+				if((int)microtime(true) >= ($minted+$expiration))
+				{
+					// This token is expired. Nuke it.
+					
+					delete_user_meta($user->ID,'patreon_access_token');
+			
+				}
+		
+			}
+			else
+			{
+				// No minted value. Even if there may be no access token created and saved, still nuke it.
+				
+				delete_user_meta($user->ID,'patreon_access_token');
+				
+			}
+	
+		}
+
+
+	}
+
+	public static function createOrLogInUserFromPatreon($user_response, $tokens, $redirect = false) {
 
 		global $wpdb;
 
-		$login_with_patreon = get_option('patreon-enable-login-with-patreon', false);
+		$login_with_patreon = get_option('patreon-enable-login-with-patreon', true);
+
 		$admins_editors_login_with_patreon = get_option('patreon-enable-allow-admins-login-with-patreon', false);
+		
+		$danger_user_list = Patreon_Login::getDangerUserList();
+		
+		// Chec if user is logged in to wp:
+		
+		// Logged in user. We just link him up and be done.
+		if(is_user_logged_in()) {
+			
+			$user = wp_get_current_user();
+						
+			self::updateExistingUser($user->ID, $user_response, $tokens);
+			wp_redirect($redirect);
+			exit;		
 
-		$email = $user_response['data']['attributes']['email'];
+		}
+		
+		////////////////////////////////////////////////
+		// If we are here, User is not logged in. Go through login or creation procedure:
+		////////////////////////////////////////////////
+		
+		$patreon_user_id = $user_response['data']['id'];
+		
+		// First see if any user was linked to this patreon user:
+		
+		global $wpdb;
+		
+		$prepared_sql = $wpdb->prepare(
+		
+						"SELECT * FROM ".$wpdb->usermeta." WHERE meta_key = 'patreon_user_id' AND meta_value = %s",
+						
+						array($patreon_user_id)
+			  
+		);
+		
+		// Now get the result : 
+	
+		$patreon_linked_accounts = $wpdb->get_results($prepared_sql,ARRAY_A);
+		
+		if(count($patreon_linked_accounts)>0)
+		{
+			////////////////////////////////////////////////
+			// We have linked users. Get the last login dates for each. The reason we did it by taking ids and iterating them is to avoid querying both patreon ids and login dates at the same time and having to use a self join query on wp_usermeta
+			////////////////////////////////////////////////
+			
+			$sort_logins = array();
+			
+			foreach($patreon_linked_accounts as $key => $value)
+			{
```

Comment:
format this foreach just like the ifs

Reply:
Formatting commit:

https://github.com/codebard/patreon-wordpress/commit/e0037a62a35a355a8897b8b6cd8c6fd4a8f93b52